tests/services/test_knowledge_svc.py

```diff
@@ -0,0 +1,112 @@
+from app.objects.secondclass.c_fact import Fact
+from app.objects.secondclass.c_relationship import Relationship
+from app.objects.secondclass.c_rule import Rule
+
+
+class TestKnowledgeService:
+
+    def test_no_duplicate_fact(self, loop, knowledge_svc):
+        loop.run_until_complete(knowledge_svc.add_fact(Fact(trait='test', value='demo', score=1,
+                                                            collected_by='thin_air', technique_id='T1234')))
+        loop.run_until_complete(knowledge_svc.add_fact(Fact(trait='test', value='demo', score=1,
+                                                            collected_by='thin_air', technique_id='T1234')))
+        facts = loop.run_until_complete(knowledge_svc.get_facts(dict(trait='test')))
+        assert len(facts) == 1
+
+    def test_no_duplicate_rules(self, loop, knowledge_svc):
+        loop.run_until_complete(knowledge_svc.add_rule(Rule(action='BLOCK', trait='a.c', match='.*')))
+        loop.run_until_complete(knowledge_svc.add_rule(Rule(action='BLOCK', trait='a.c', match='.*')))
+        rules = loop.run_until_complete(knowledge_svc.get_rules(dict(trait='a.c')))
+        assert len(rules) == 1
+
+    def test_no_duplicate_relationship(self, loop, knowledge_svc):
+        dummy = Fact(trait='test', value='demo', score=1, collected_by='thin_air', technique_id='T1234')
+        loop.run_until_complete(knowledge_svc.add_relationship(Relationship(source=dummy, edge='potato', target=dummy)))
+        loop.run_until_complete(knowledge_svc.add_relationship(Relationship(source=dummy, edge='potato', target=dummy)))
+        relationships = loop.run_until_complete(knowledge_svc.get_relationships(dict(edge='potato')))
+        assert len(relationships) == 1
+
+    def test_remove_fact(self, loop, knowledge_svc):
+        loop.run_until_complete(knowledge_svc.add_fact(Fact(trait='rtest', value='rdemo', score=1,
+                                                            collected_by='thin_air', technique_id='T1234'),
+                                                       constraints=dict(test_field='test_value')))
+        loop.run_until_complete(knowledge_svc.add_fact(Fact(trait='ktest', value='rdemo', score=1,
+                                                            collected_by='thin_air', technique_id='T1234'),
+                                                       constraints=dict(test_field='test_value2')))
+        loop.run_until_complete(knowledge_svc.delete_fact(dict(trait='rtest')))
+        facts = loop.run_until_complete(knowledge_svc.get_facts(dict(value='rdemo')))
+        assert len(facts) == 1
```

Comment:
Should the removal tests also check that the constraint was deleted, too?

Reply:
Yeah, that makes sense. I'll add it.